### DISCUSSION TITLE
Script Suggestion - immich.app

### DISCUSSION THREAD

**Discussion by ghost** [tteck/Proxmox#1864] [Ideas]
https://github.com/immich-app/immich

I think this would make a good addition to the current library. Any thoughts?

**Comment by vednolacni** [comment]
+1

**Comment by Vakrehus** [comment]
+1

I've had immich.app running on Docker for the whole family for a few months now and it's a great performant alternative to Google Photos.
Whether immich.app runs even better without Docker, I can not judge, but would be very happy if there would be a script for it.

**Comment by mushis** [comment]
+1

**Comment by IcedOrange00** [comment]
+1

**Comment by tteck** [comment]
![Screenshot 2023-11-20 8 13 35 AM](https://github.com/tteck/Proxmox/assets/21342670/ba9edb66-11f5-4912-955b-f732a59cd8d2)

**Reply by vednolacni** [reply]
greatly appreciated!

**Reply by ssldany897** [reply]
Is there any difference between running it with dockge or directly as a LXC

**Reply by tteck** [reply]
There is no "_directly as a LXC_"

**Reply by johndoe0815** [reply]
Does the bash command then also update the Immich application to the latest release or is this then the usual manual process like usually also the inital setup?

**Reply by tteck** [reply]
I don't understand the question. If you're talking about the Dockge LXC, it adds the Immich compose file from Immich. All that is needed is to start it in Dockge.

**Reply by johndoe0815** [reply]
Then let me rephrase it: The script allows to install Immich (by adding the current standard compose file as it seems).
In case Immich issues an update, that sometimes also requires a change in the docker compose file, is anything of that done be re-executing the initial script? Otherwise it basically only automates an initial copy & paste from the immich docker compose file into dockge.
I am just trying to understand if this approach here offers any advantage over my current way (using your docker LXC with Portainer and manually deploying Immich in there).

**Reply by tteck** [reply]
It's just standard Docker Compose with a webui. When there's an Immich update, hit the update button.

![Screenshot 2023-11-28 10 15 48 AM](https://github.com/tteck/Proxmox/assets/21342670/55719861-daae-40a0-b4a9-4146d530c592)

edit: I'd suggest testing it and deciding for yourself which you prefer.

**Comment by mastan30** [comment]
Hi Anyone trying to install immich in Proxmox, here's a guide I created using docker lxc unpriviliged container and docker compose of immich.



# Immich Docker Install

1. To install immich, the recommended way to install is using a docker compose, we can't install it on a bare metal debian os. So we will install a docker LXC using the proxmox script and install docker compose.Then we will create the docker-compose file for installing immich.

2. To install docker LXC we can used the following Proxmox script from the beloved tteck github:

		

		Docker LXC
		Options to Install Portainer and/or Docker Compose V2
		Docker is an open-source project for automating the deployment of applications as portable, self-sufficient containers.

		🛈 If the LXC is created Privileged, the script will automatically set up USB passthrough.

		To create a new Proxmox VE Docker LXC, run the command below in the Proxmox VE Shell.

		Copy
		bash -c "$(wget -qLO - https://github.com/tteck/Proxmox/raw/main/ct/docker.sh)"
		⚡ Default Settings: 2GB RAM - 4GB Storage - 2vCPU ⚡
		As an alternative option, you can use Alpine Linux and the Docker package to create a Docker LXC container with faster creation time and minimal system resource usage.

		To create a new Proxmox VE Alpine-Docker LXC, run the command below in the Proxmox VE Shell.

		Copy
		bash -c "$(wget -qO - https://github.com/tteck/Proxmox/raw/main/ct/alpine-docker.sh)"
		⚡ Default Settings: 1GB RAM - 2GB Storage - 1vCPU ⚡
		⚠ Run Compose V2 by replacing the hyphen (-) with a space, using docker compose, instead of docker-compose.

		Portainer Interface: IP:9443


3. Once the container is installed, let's mount a mountpoint of SMB share where the images are uploaded.

4. In order to write the files into the SMB share, there's a catch here. The LXC we created using the script is an unpriviliged container which is the safest way to build container but the issue is we can't write anything into the shared drived from Proxmox host.

5. Instead of mounting the SMB share via Proxmox UI, we use fstab to mount the SMB share. Here are the steps to mount

	```
	  vi /etc/fstab
	```
	
	Then add the following line replacing the uid and gid for which you want the access to the proxmox container
	
	```
	  //192.168.X.XX/Path/to/share/ /path/to/mount/       cifs    noperm,iocharset=utf8,rw,credentials=/root/.storage_credentials,uid=1000,gid=1000,file_mode=0660,dir_mode=0770 0 0
	```
     eg: //192.168.1.19/Photos/immichtest/ /mnt/pve/immichtest/       cifs    noperm,iocharset=utf8,rw,credentials=/root/.storage_credentials,uid=1000,gid=1000,file_mode=0660,dir_mode=0770 0 0
	 
	```
      vi /root/.storage_credentials
    ```	
	 In this file add the username and password as below:
	 
	```
	username=username
	password=password
	```
	
	```
	mount -a
	```
   This will mount the SMB share into the Proxmox host, now check the mount directory permissions by doing ls -al and check if the mount has the permissions of assigned uid and gid. If not do a reboot of the host and it takes care of it.
   
6. Shutdown the container and mount the SMB share to the proxmox LXC by following commands:

   ```
    pct set 100 -mp0 /mnt/bindmounts/shared,mp=/shared
   
   ```
    eg: pct set 100 -mp0 /mnt/pve/immichtest/ ,mp=/immichtest/   - 100 is the lxc id
   
7. Start the container and check the if the mount point is assigned with right uid and gid inside the container.

8. Now in the LXC create a folder using following commands

   ```
     mkdir ./immich/
	 cd ./immich/
   ```
   
   create the docker-compose file and also the .env file in the directory as per the immich guide
   
   ```
      https://immich.app/docs/install/docker-compose
   ```
   
   Run the docker compose and follow the immichi guide to setup your albums and libraries as required.
   
   
9. Optinally in order to add hardware acceleration, I used NVIDIA acceleration. In order to get it working here are the steps to follow:
   
   Next step is to passthrough the nvidia graphics card to the lxc, In order to do the passthrough follow the below guide.
   
   > [Nvidia graphics passthrough](https://gist.github.com/egg82/90164a31db6b71d36fa4f4056bbee2eb_
   > [Nivdia graphics passthrough video](https://www.youtube.com/watch?v=-Us8KPOhOCY)
   
   I have already installed NVIDIA drivers for jellyfin lxc in proxmox host, so I just installed the nvidia drivers in container using below commands:
   
   ```
    wget https://us.download.nvidia.com/XFree86/Linux-x86_64/535.146.02/NVIDIA-Linux-x86_64-{Latest-version}.run
   ```
    you can get the NVIDIA latest driver url from here : https://www.nvidia.com/Download/index.aspx
	
   once downloaded install the drivers (The --no-kernel-module module is used as already there's kernel installed ):
   ```
   chmod +x NVIDIA-Linux-x86_64-{Latest-version}.run
   
   ./NVIDIA-Linux-x86_64-{Latest-version}.run ----no-kernel-module
   ```
   
   once installed make sure the drivers are installed by running:
   ```
   nvidia-smi -l 10
   ```
   
10. Now we need to pass through the hardware acceleration to docker container by following the immich guide:
    
	```
	https://immich.app/docs/features/hardware-transcoding
    ```
	
	In order to have the driver passed to container, we need to have nvidia container toolkit installed using the following commands in the container:
	(https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/latest/install-guide.html)
	
	```
	curl -fsSL https://nvidia.github.io/libnvidia-container/gpgkey | sudo gpg --dearmor -o /usr/share/keyrings/nvidia-container-toolkit-keyring.gpg \
    && curl -s -L https://nvidia.github.io/libnvidia-container/stable/deb/nvidia-container-toolkit.list | \
    sed 's#deb https://#deb [signed-by=/usr/share/keyrings/nvidia-container-toolkit-keyring.gpg] https://#g' | \
    sudo tee /etc/apt/sources.list.d/nvidia-container-toolkit.list
	
	
	
	sed -i -e '/experimental/ s/^#//g' /etc/apt/sources.list.d/nvidia-container-toolkit.list
	
	
	
	sudo apt-get update
	
	
	sudo apt-get install -y nvidia-container-toolkit
	
	
	```
	
	Also one other step I needed to do to make sure the docker compose works with hardware-transcoding is to modify nvidia config using below commands in container:
	
	Edit /etc/nvidia-container-runtime/config.toml and disable cgroups:
	```
      
	  vi /etc/nvidia-container-runtime/config.toml
	  
	  set no-cgroups = true in the config.toml file, save and reboot the container. Now run the docker compose for immich.

	```
	
	
	
	References:
	
	https://immich.app/docs/install/script
	
	https://forum.proxmox.com/threads/how-to-automatically-mount-a-smb-share-on-an-lxc.118220/

    https://immich.app/docs/features/hardware-transcoding
	
	https://www.reddit.com/r/Proxmox/comments/s0ud5y/comment/jl4lef2/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button
	
	<blockquote class="reddit-embed-bq" data-embed-height="348"><a href="https://www.reddit.com/r/Proxmox/comments/s0ud5y/comment/jl4lef2/">Comment</a><br> by<a href="https://www.reddit.com/user/Masterofchicken/">u/Masterofchicken</a> from discussion<a href="https://www.reddit.com/r/Proxmox/comments/s0ud5y/cgroups2_problem_with_nvidiacontainercli/"><no value=""></no></a><br> in<a href="https://www.reddit.com/r/Proxmox/">Proxmox</a></blockquote><script async="" src="https://embed.reddit.com/widgets.js" charset="UTF-8"></script>
